# AIA Machine Learning

### Sepsis Prediction Overview

Sepsis is the leading cause of mortality in US hospitals and the most expensive condition associated with in-hospital stays, accounting for 6.2% (nearly $24 billion) of total hospital stay costs. In particular, septic shock, the most advanced complication of sepsis, due to severe abnormalities of circulation and/or cellular metabolism, reaches a mortality rate as high as 50%, and the annual incidence keeps rising. It is estimated that as much as 80% of sepsis deaths could be prevented with early diagnosis and intervention. Indeed, prior studies have demonstrated that early diagnosis and treatment of septic shock can significantly decrease patient mortality and shorten their hospital stay. In this capstone, the task is to build a machine learning model for accurate early diagnosis of septic shock.

*This Jupyter Notebook uses existing Python files for the below RTP mining activities:*
https://github.com/farzanehkh/RTP-mining-ICHI18

In [1]:
import sys
import os
sys.stderr = open(os.devnull, 'w')
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'  # 0 = all logs, 1 = filter INFO, 2 = filter WARNING, 3 = filter ERROR

I0000 00:00:1760859310.711422   43625 gpu_device.cc:2020] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 9558 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 4070, pci bus id: 0000:01:00.0, compute capability: 8.9


In [2]:
import RTP
import warnings
import numpy as np
import pandas as pd
from collections import defaultdict
from concurrent.futures import ThreadPoolExecutor

import tensorflow as tf
tf.get_logger().setLevel('ERROR')
from tensorflow import keras
import tensorflow.random as tf_random

from sklearn.model_selection import train_test_split, KFold
from sklearn.metrics import roc_curve, roc_auc_score
from sklearn.metrics import recall_score, accuracy_score
from sklearn.metrics import precision_score, f1_score

warnings.filterwarnings("ignore", category=UserWarning, module='tensorflow')
tf_random.set_seed(2022)
random_state = 2022
print("Num GPUs Available: ", len(tf.config.list_physical_devices('GPU')))

Num GPUs Available:  1


In [3]:
# Load raw data from CSV files
shock = pd.read_csv('mimic/MIMIC_III_shock.csv')
nonshock = pd.read_csv('mimic/MIMIC_III_nonshock.csv')

# Select relevant clinical features
positive = shock[['VisitIdentifier', 'MinutesFromArrival', 'SystolicBP', 
                  'HeartRate', 'RespiratoryRate', 'Temperature', 'WBC', 'ShockTime']].copy()
negative = nonshock[['VisitIdentifier', 'MinutesFromArrival', 'SystolicBP', 
                     'HeartRate', 'RespiratoryRate', 'Temperature', 'WBC', 'LastMinute']].copy()

# Ensure consistent data types
negative['LastMinute'] = negative['LastMinute'].astype('int64')

# Rename time-to-diagnosis columns to a common name
positive = positive.rename(columns={'ShockTime': 'DiagnosisMinutes'})
negative = negative.rename(columns={'LastMinute': 'DiagnosisMinutes'})

# Filter out visits that ended less than 24 hours after arrival
positive = positive.query('DiagnosisMinutes - MinutesFromArrival >= 1440')
negative = negative.query('DiagnosisMinutes - MinutesFromArrival >= 1440')

# Drop time-related columns
positive = positive.drop(columns=['DiagnosisMinutes', 'MinutesFromArrival'])
negative = negative.drop(columns=['DiagnosisMinutes', 'MinutesFromArrival'])

# Add binary diagnosis label
positive['Diagnosis'] = 1
negative['Diagnosis'] = 0

# Combine both datasets
visits = pd.concat([positive, negative], ignore_index=True)

# Print dataset shape
print(f"Combined dataset shape: {visits.shape}")

Combined dataset shape: (84858, 7)


In [4]:
# Separate the features and labels
X, y = [], []
for visit in visits['VisitIdentifier'].unique():
    df_visit = visits[visits['VisitIdentifier'] == visit]
    y.append(int(df_visit.tail(1)['Diagnosis']))
    X.append(df_visit[['SystolicBP', 'HeartRate', 'RespiratoryRate', 'Temperature', 'WBC']].values.tolist())

# Pad sequences to match the longest visit
max_len = 1000  # or 500
X_trimmed = [seq[:max_len] for seq in X]
X_padded = keras.preprocessing.sequence.pad_sequences(X_trimmed, dtype='float64', padding="post", value=-10.0)

# X_padded = keras.preprocessing.sequence.pad_sequences(X, dtype='float64', padding="post", value=-10.0)

# Convert to numpy arrays
X_padded = np.asarray(X_padded)
y = np.asarray(y)

# Split into train, validation, and test sets
X_train, X_temp, y_train, y_temp = train_test_split(X_padded, y, test_size=0.30, 
                                                    random_state=random_state, stratify=y)
X_valid, X_test, y_valid, y_test = train_test_split(X_temp, y_temp, test_size=0.67, 
                                                    random_state=random_state, stratify=y_temp)

In [5]:
def get_metrics(batch, df):
    # Define and compile a fresh model
    model = keras.Sequential([
        keras.layers.Masking(mask_value=-10.0, input_shape=X_train.shape[-2:]),
        keras.layers.LSTM(200),
        keras.layers.Dense(1, activation='sigmoid')
    ])
    opt = keras.optimizers.Adam(learning_rate=0.01)
    model.compile(optimizer=opt, loss='binary_crossentropy')

    # Train the model
    model.fit(X_train, y_train, epochs=10, batch_size=batch, validation_data=(X_valid, y_valid), verbose=1)

    # Predict and evaluate
    y_preds = model.predict(X_test).ravel()
    preds = np.where(y_preds > 0.5, 1, 0)

    auc_score = roc_auc_score(y_test, y_preds)
    accuracy = accuracy_score(y_test, preds)
    precision = precision_score(y_test, preds)
    recall = recall_score(y_test, preds)
    f_measure = f1_score(y_test, preds)

    # Record metrics
    df.loc[len(df)] = ['LSTM', batch, accuracy, precision, recall, f_measure, auc_score]
    return df

In [6]:
# Create a dataframe to store metric scores
lstm_metrics = pd.DataFrame(columns=['Method', 'batch', 'accuracy', 'precision', 'recall', 'f1 score', 'auc roc'])

# Train and evaluate the model for each batch size
for batch_size in [64, 128, 256]:
    lstm_metrics = get_metrics(batch_size, lstm_metrics)
    lstm_metrics.to_csv('metrics/sepsis_lstm_metrics.csv', index=False)

Epoch 1/10
9/9 ━━━━━━━━━━━━━━━━━━━━ 1s 55ms/step - loss: 0.7910 - val_loss: 0.6988
Epoch 2/10
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 30ms/step - loss: 0.6938 - val_loss: 0.6936
Epoch 3/10
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 30ms/step - loss: 0.6927 - val_loss: 0.6944
Epoch 4/10
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 30ms/step - loss: 0.6896 - val_loss: 0.6927
Epoch 5/10
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 31ms/step - loss: 0.6875 - val_loss: 0.6952
Epoch 6/10
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 31ms/step - loss: 0.6876 - val_loss: 0.6929
Epoch 7/10
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 30ms/step - loss: 0.6904 - val_loss: 0.6990
Epoch 8/10
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 31ms/step - loss: 0.6892 - val_loss: 0.6926
Epoch 9/10
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 39ms/step - loss: 0.6826 - val_loss: 0.6972
Epoch 10/10
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 41ms/step - loss: 0.6830 - val_loss: 0.6913
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 18ms/step
Epoch 1/10
5/5 ━━━━━━━━━━━━━━━━━━━━ 1s 82ms/step - loss: 0.9025 - val_loss: 0.7601
Epoch 2/10
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 65ms/step

In [7]:
# Define column structure for results
columns = ['Method', 'model', 'max gap', 'sup pos', 'sup neg', 
           'accuracy', 'precision', 'recall', 'f1 score', 'auc roc']

# Define parameter grid
param_grid = [(gap, pos_sup, neg_sup, clf) 
              for clf in ['svm', 'lr']
              for gap in range(4, 11)
              for pos_sup in [0.1, 0.15, 0.2, 0.25, 0.3]
              for neg_sup in [0.1, 0.15, 0.2, 0.25, 0.3]]

# Define single trial runner
def run_trial(params):
    gap, pos_sup, neg_sup, clf = params
    metrics = RTP.main(gap, pos_sup, neg_sup, clf)
    return ['RTP', clf, gap, pos_sup, neg_sup] + metrics

# Run trials in parallel
with ThreadPoolExecutor() as executor:
    results = list(executor.map(run_trial, param_grid))

# Convert results to DataFrame
rtp_metrics = pd.DataFrame(results, columns=columns)

# Save to CSV
rtp_metrics.to_csv('metrics/sepsis_rtp_metrics.csv', index=False)

In [8]:
# Load and round RTP results
rtp_metrics = pd.read_csv('metrics/sepsis_rtp_metrics.csv').round(3)

# Load and round LSTM results
lstm_metrics = pd.read_csv('metrics/sepsis_lstm_metrics.csv').round(3)

# Display best RTP scores across all metrics
print("Best RTP Model Scores:")
for metric in ['accuracy', 'precision', 'recall', 'f1 score', 'auc roc']:
    print(f"\nHighest RTP {metric}:")
    display(rtp_metrics[rtp_metrics[metric] == rtp_metrics[metric].max()])

# Display best LSTM scores across all metrics
print("\nBest LSTM Model Scores:")
for metric in ['accuracy', 'precision', 'recall', 'f1 score', 'auc roc']:
    print(f"\nHighest LSTM {metric}:")
    display(lstm_metrics[lstm_metrics[metric] == lstm_metrics[metric].max()])


Best RTP Model Scores:

Highest RTP accuracy:


,Method,model,max gap,sup pos,sup neg,accuracy,precision,recall,f1 score,auc roc
25,RTP,svm,5,0.1,0.1,0.842,0.77,0.975,0.86,0.883



Highest RTP precision:


,Method,model,max gap,sup pos,sup neg,accuracy,precision,recall,f1 score,auc roc
222,RTP,lr,5,0.3,0.20,0.838,0.831,0.848,0.84,0.916
223,RTP,lr,5,0.3,0.25,0.838,0.831,0.848,0.84,0.917
224,RTP,lr,5,0.3,0.30,0.838,0.831,0.848,0.84,0.917
247,RTP,lr,6,0.3,0.20,0.838,0.831,0.848,0.84,0.916
248,RTP,lr,6,0.3,0.25,0.838,0.831,0.848,0.84,0.916
249,RTP,lr,6,0.3,0.30,0.838,0.831,0.848,0.84,0.916
272,RTP,lr,7,0.3,0.20,0.838,0.831,0.848,0.84,0.916
273,RTP,lr,7,0.3,0.25,0.838,0.831,0.848,0.84,0.916
274,RTP,lr,7,0.3,0.30,0.838,0.831,0.848,0.84,0.916
297,RTP,lr,8,0.3,0.20,0.838,0.831,0.848,0.84,0.916



Highest RTP recall:


,Method,model,max gap,sup pos,sup neg,accuracy,precision,recall,f1 score,auc roc
50,RTP,svm,6,0.1,0.1,0.841,0.767,0.977,0.86,0.882
75,RTP,svm,7,0.1,0.1,0.841,0.767,0.977,0.86,0.883
100,RTP,svm,8,0.1,0.1,0.841,0.767,0.977,0.86,0.883



Highest RTP f1 score:


,Method,model,max gap,sup pos,sup neg,accuracy,precision,recall,f1 score,auc roc
25,RTP,svm,5,0.1,0.1,0.842,0.770,0.975,0.86,0.883
50,RTP,svm,6,0.1,0.1,0.841,0.767,0.977,0.86,0.882
75,RTP,svm,7,0.1,0.1,0.841,0.767,0.977,0.86,0.883
100,RTP,svm,8,0.1,0.1,0.841,0.767,0.977,0.86,0.883



Highest RTP auc roc:


,Method,model,max gap,sup pos,sup neg,accuracy,precision,recall,f1 score,auc roc
183,RTP,lr,4,0.15,0.25,0.834,0.824,0.851,0.837,0.917
184,RTP,lr,4,0.15,0.30,0.834,0.824,0.851,0.837,0.917
207,RTP,lr,5,0.15,0.20,0.837,0.828,0.851,0.839,0.917
208,RTP,lr,5,0.15,0.25,0.837,0.828,0.851,0.839,0.917
209,RTP,lr,5,0.15,0.30,0.837,0.828,0.851,0.839,0.917
213,RTP,lr,5,0.20,0.25,0.834,0.827,0.846,0.836,0.917
214,RTP,lr,5,0.20,0.30,0.834,0.827,0.846,0.836,0.917
221,RTP,lr,5,0.30,0.15,0.832,0.828,0.838,0.833,0.917
223,RTP,lr,5,0.30,0.25,0.838,0.831,0.848,0.840,0.917
224,RTP,lr,5,0.30,0.30,0.838,0.831,0.848,0.840,0.917



Best LSTM Model Scores:

Highest LSTM accuracy:


,Method,batch,accuracy,precision,recall,f1 score,auc roc
1,LSTM,128,0.683,0.708,0.63,0.667,0.733



Highest LSTM precision:


,Method,batch,accuracy,precision,recall,f1 score,auc roc
1,LSTM,128,0.683,0.708,0.63,0.667,0.733



Highest LSTM recall:


,Method,batch,accuracy,precision,recall,f1 score,auc roc
0,LSTM,64,0.54,0.53,0.753,0.622,0.619



Highest LSTM f1 score:


,Method,batch,accuracy,precision,recall,f1 score,auc roc
1,LSTM,128,0.683,0.708,0.630,0.667,0.733
2,LSTM,256,0.652,0.644,0.691,0.667,0.690



Highest LSTM auc roc:


,Method,batch,accuracy,precision,recall,f1 score,auc roc
1,LSTM,128,0.683,0.708,0.63,0.667,0.733


In [9]:
# Best Settings per Model
print("Best Hyperparameter Settings\n")

# LSTM
best_batch = []
print("LSTM Settings")
print("-" * 30)
for metric in ['accuracy', 'precision', 'recall', 'f1 score', 'auc roc']:
    best_param = lstm_metrics[lstm_metrics[metric] == lstm_metrics[metric].max()]['batch'].mode()[0]
    best_batch.append(best_param)
    print(f"Best {metric}: batch size = {best_param}")
print(f"\nBest overall batch size: {max(best_batch, key=best_batch.count)}\n")

# RTP-SVM and RTP-LR
for model in ['svm', 'lr']:
    best_params = defaultdict(list)
    print(f"RTP-{model.upper()} Settings")
    print("-" * 30)
    rtp_model = rtp_metrics[rtp_metrics['model'] == model]
    for metric in ['accuracy', 'precision', 'recall', 'f1 score', 'auc roc']:
        for param in ['max gap', 'sup pos', 'sup neg']:
            best_param = rtp_model[rtp_model[metric] == rtp_model[metric].max()][param].mode()[0]
            best_params[param].append(best_param)
            print(f"Best {metric}: {param} = {best_param}")
        print()
    print("Best overall settings:")
    for k, v in best_params.items():
        print(f"{k}: {max(v, key=v.count)}")
    print()

# Highest Metric Scores Across All Models
highest_metrics = pd.DataFrame(columns=['Method', 'Metric', 'Score'])

for metric in ['accuracy', 'precision', 'recall', 'f1 score', 'auc roc']:
    rtp_max = rtp_metrics[metric].max()
    lstm_max = lstm_metrics[metric].max()
    if rtp_max >= lstm_max:
        model = rtp_metrics[rtp_metrics[metric] == rtp_metrics[metric].max()]['model'].iloc[0]
        highest_metrics.loc[len(highest_metrics)] = [f"RTP-{model}", metric, rtp_max]
    else:
        batch = lstm_metrics[lstm_metrics[metric] == lstm_metrics[metric].max()]['batch'].iloc[0]
        highest_metrics.loc[len(highest_metrics)] = [f"LSTM-{batch}", metric, lstm_max]

# Final Summary
print("Results and Final Report\n")
print("Machine Learning Methods Compared:")
print("* Recent Temporal Pattern (RTP)")
print("  - Support Vector Machine (SVM)")
print("  - Logistic Regression (LR)")
print("* Long Short-Term Memory (LSTM)\n")

print("Methods with the Highest Metric Scores:")
print(highest_metrics.round(3).to_string(index=False))

Best Hyperparameter Settings

LSTM Settings
------------------------------
Best accuracy: batch size = 128
Best precision: batch size = 128
Best recall: batch size = 64
Best f1 score: batch size = 128
Best auc roc: batch size = 128

Best overall batch size: 128

RTP-SVM Settings
------------------------------
Best accuracy: max gap = 5
Best accuracy: sup pos = 0.1
Best accuracy: sup neg = 0.1

Best precision: max gap = 5
Best precision: sup pos = 0.1
Best precision: sup neg = 0.15

Best recall: max gap = 6
Best recall: sup pos = 0.1
Best recall: sup neg = 0.1

Best f1 score: max gap = 5
Best f1 score: sup pos = 0.1
Best f1 score: sup neg = 0.1

Best auc roc: max gap = 4
Best auc roc: sup pos = 0.15
Best auc roc: sup neg = 0.25

Best overall settings:
max gap: 5
sup pos: 0.1
sup neg: 0.1

RTP-LR Settings
------------------------------
Best accuracy: max gap = 5
Best accuracy: sup pos = 0.3
Best accuracy: sup neg = 0.2

Best precision: max gap = 5
Best precision: sup pos = 0.3
Best preci

#### Best Overall Settings and Parameters
* RTP-SVM
  * max gap: 5
  * sup pos: 0.1
  * sup neg: 0.1
* RTP-LR
  * max gap: 5
  * sup pos: 0.3
  * sup neg: 0.2
* LSTM
  * batch: 128

#### Metrics Evaluation
* True Positive Rate (TP) - *The probability that an actual positive correctly tests positive.*
* True Negative Rate (TN) - *The probability that an actual negative correctly tests negative.*
* False Positive Rate (FP) - *The probability that an actual negative wrongly tests positive.*
* False Negative Rate (FN) - *The probability that an actual positive wrongly tests negative.*

##### Accuracy: (TP + TN) / (TP + FP + FN + TN) 
* RTP-SVM correctly predicted a combined 84% of positive and negative cases. Not very informative for sepsis diagnosis due to the fact that false negatives (sepsis patients misidentified as non-sepsis patients) have a much higher cost than false positives (non-sepsis patients misidentified as sepsis patients), and Accuracy doesn't distinguish between the two.

##### Precision: TP / (TP + FP) 
* RTP-LR correctly classified 83% of its positive predictions. This tells us that our sepsis positive predictions are reasonably reliable as long as the counter-measures for sepsis patients are not particularly invasive or resource-intensive, since misidentified non-sepsis patients would receive the same treatment as correctly identified sepsis patients.

##### Recall: TP / (TP + FN) 
* RTP-SVM correctly classified 97% of all positive cases. This is the most important metric for sepsis diagnosis since the accurate diagnosis of sepsis-positive patients is the priority here. As such, with a 97% Recall rate, we are reasonably certain of our ability to intercede on behalf of sepsis patients when necessary.

##### F1-score: 2 x (Recall x Precision) / (Recall + Precision) 
* RTP-SVM has a fairly high 0.86 weighted average score between Recall and Precision. This is the ability to both identify sepsis cases (Recall) and be accurate with the cases it predicts as sepsis (Precision). In other words, the confidence we have in our ability to intervene on behalf of and spend resources on those patients and only those patients who truly need intervention.

##### AUC-ROC: Sensitivity vs 1 − Specificity or TP Rate vs FP Rate
* RTP-LR has the highest AUC-ROC score of 0.91, indicating its superior performance over LSTM at distinguishing between patients with sepsis and patients without sepsis at various thresholds. 

#### Analysis:

Having completed a rigorous exploration of the sepsis EHR data using the methods RTP-SVM, RTP-LR, and LSTM, we have found that RTP outperformed LSTM on all of the relevant metrics. Furthermore, the fitted models, SVM and LR, used in conjunction with RTP, each showed promise by achieving high scores in the various metrics. Of these, it is our conclusion that RTP-SVM is the most promising method for clinical use due to the fact that it claims the high scores for the most relevant metrics for sepsis prediction, the Recall and F1-score. Using these scores, we can be confident in our ability to intervene on behalf of sepsis patients while keeping unnecessary intervention levels to an acceptable minimum.

#### Proposal:

After reviewing Dr. Chi’s paper, "Recent Temporal Pattern Mining for Septic Shock Early Prediction", we found that their work cited an almost identical RTP F1-score of 0.868 using an observation window just 4 hours before diagnosis, versus our F1-score of 0.860 using the same method 24 hours before diagnosis. That is a full 20 hours increase in advance notice with little drop-off in predictive ability. With such a minimal decrease in accuracy over such an extended period of time, we feel it would be reasonable to set a minimum acceptable F1-score and walk back the observation window in 4-hour intervals until the minimum F1-score is met. This will help to identify patients with sepsis that much earlier and before the disease has a chance to progress.
